File: desktop/MRSS.Native/Services/StandardTranslationClient.cs
using System.Globalization;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using MRSS.Native.Models;

namespace MRSS.Native.Services;

public sealed class StandardTranslationClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);
    private static readonly Regex HtmlTagRegex = new("(?s)<[^>]+>", RegexOptions.Compiled);
    private readonly HttpClient _httpClient = new() { Timeout = TimeSpan.FromSeconds(45) };

    public async Task<ArticleTranslation> TranslateAsync(StandardTranslationSettings settings, TranslationJob job, CancellationToken cancellationToken = default)
    {
        var title = await TranslateTextPreservingFormatAsync(settings, job.Title, job.TargetLanguage, cancellationToken);
        var content = await TranslateTextPreservingFormatAsync(settings, job.Content, job.TargetLanguage, cancellationToken);
        return new ArticleTranslation { Title = title, Content = content };
    }

    private async Task<string> TranslateTextPreservingFormatAsync(StandardTranslationSettings settings, string? source, string targetLanguage, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(source))
        {
            return source ?? "";
        }

        if (!LooksLikeHtml(source))
        {
            return await TranslatePlainAsync(settings, source, targetLanguage, cancellationToken);
        }

        var builder = new StringBuilder(source.Length);
        var index = 0;
        foreach (Match match in HtmlTagRegex.Matches(source))
        {
            if (match.Index > index)
            {
                var text = source[index..match.Index];
                builder.Append(await TranslateTextPartAsync(settings, text, targetLanguage, cancellationToken));
            }

            builder.Append(match.Value);
            index = match.Index + match.Length;
        }

        if (index < source.Length)
        {
            builder.Append(await TranslateTextPartAsync(settings, source[index..], targetLanguage, cancellationToken));
        }

        return builder.ToString();
    }

    private async Task<string> TranslateTextPartAsync(StandardTranslationSettings settings, string text, string targetLanguage, CancellationToken cancellationToken)
    {
        return string.IsNullOrWhiteSpace(text)
            ? text
            : await TranslatePlainAsync(settings, text, targetLanguage, cancellationToken);
    }

    private Task<string> TranslatePlainAsync(StandardTranslationSettings settings, string source, string targetLanguage, CancellationToken cancellationToken)
    {
        return NormalizeProvider(settings.Provider) switch
        {
            "baidu" => TranslateBaiduAsync(settings, source, targetLanguage, cancellationToken),
            "tencent" => TranslateTencentAsync(settings, source, targetLanguage, cancellationToken),
            "google" => TranslateGoogleAsync(settings, source, targetLanguage, cancellationToken),
            _ => TranslateMicrosoftAsync(settings, source, targetLanguage, cancellationToken)
        };
    }

    private async Task<string> TranslateBaiduAsync(StandardTranslationSettings settings, string source, string targetLanguage, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(settings.BaiduAppId) || string.IsNullOrWhiteSpace(settings.BaiduSecret))
        {
            throw new InvalidOperationException("请先填写百度翻译 App ID 和密钥。");
        }

        var salt = RandomNumberGenerator.GetInt32(100000, 999999).ToString(CultureInfo.InvariantCulture);
        var sign = Md5Hex(settings.BaiduAppId + source + salt + settings.BaiduSecret);
        var form = new Dictionary<string, string>
        {
            ["q"] = source,
            ["from"] = "auto",
            ["to"] = LanguageForBaidu(targetLanguage),
            ["appid"] = settings.BaiduAppId.Trim(),
            ["salt"] = salt,
            ["sign"] = sign
        };
        using var response = await _httpClient.PostAsync("https://fanyi-api.baidu.com/api/trans/vip/translate", new FormUrlEncodedContent(form), cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        EnsureSuccess(response, text);
        using var document = JsonDocument.Parse(text);
        if (document.RootElement.TryGetProperty("error_msg", out var error))
        {
            throw new InvalidOperationException(error.GetString() ?? "百度翻译失败。");
        }

        return string.Join("", document.RootElement.GetProperty("trans_result").EnumerateArray().Select(item => item.GetProperty("dst").GetString() ?? ""));
    }

    private async Task<string> TranslateGoogleAsync(StandardTranslationSettings settings, string source, string targetLanguage, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(settings.GoogleApiKey))
        {
            throw new InvalidOperationException("请先填写 Google 翻译 API Key。");
        }

        var payload = new
        {
            q = source,
            target = LanguageForGoogle(targetLanguage),
            format = "text"
        };
        var uri = $"https://translation.googleapis.com/language/translate/v2?key={Uri.EscapeDataString(settings.GoogleApiKey.Trim())}";
        using var response = await _httpClient.PostAsync(uri, JsonContent(payload), cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        EnsureSuccess(response, text);
        using var document = JsonDocument.Parse(text);
        return document.RootElement.GetProperty("data").GetProperty("translations")[0].GetProperty("translatedText").GetString() ?? "";
    }

    private async Task<string> TranslateMicrosoftAsync(StandardTranslationSettings settings, string source, string targetLanguage, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(settings.MicrosoftKey))
        {
            throw new InvalidOperationException("请先填写微软翻译 Key。");
        }

        var payload = new[] { new { Text = source } };
        var uri = $"https://api.cognitive.microsofttranslator.com/translate?api-version=3.0&to={Uri.EscapeDataString(LanguageForMicrosoft(targetLanguage))}";
        using var request = new HttpRequestMessage(HttpMethod.Post, uri);
        request.Content = JsonContent(payload);
        request.Headers.TryAddWithoutValidation("Ocp-Apim-Subscription-Key", settings.MicrosoftKey.Trim());
        if (!string.IsNullOrWhiteSpace(settings.MicrosoftRegion) && !settings.MicrosoftRegion.Equals("global", StringComparison.OrdinalIgnoreCase))
        {
            request.Headers.TryAddWithoutValidation("Ocp-Apim-Subscription-Region", settings.MicrosoftRegion.Trim());
        }

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        EnsureSuccess(response, text);
        using var document = JsonDocument.Parse(text);
        return document.RootElement[0].GetProperty("translations")[0].GetProperty("text").GetString() ?? "";
    }

    private async Task<string> TranslateTencentAsync(StandardTranslationSettings settings, string source, string targetLanguage, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(settings.TencentSecretId) || string.IsNullOrWhiteSpace(settings.TencentSecretKey))
        {
            throw new InvalidOperationException("请先填写腾讯翻译 SecretId 和 SecretKey。");
        }

        const string host = "tmt.tencentcloudapi.com";
        const string service = "tmt";
        const string action = "TextTranslate";
        const string version = "2018-03-21";
        var region = string.IsNullOrWhiteSpace(settings.TencentRegion) ? "ap-beijing" : settings.TencentRegion.Trim();
        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var date = DateTimeOffset.FromUnixTimeSeconds(timestamp).UtcDateTime.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        var payload = JsonSerializer.Serialize(new
        {
            SourceText = source,
            Source = "auto",
            Target = LanguageForTencent(targetLanguage),
            ProjectId = 0
        }, JsonOptions);

        var hashedPayload = Sha256Hex(payload);
        var canonicalHeaders = $"content-type:application/json; charset=utf-8\nhost:{host}\n";
        const string signedHeaders = "content-type;host";
        var canonicalRequest = $"POST\n/\n\n{canonicalHeaders}\n{signedHeaders}\n{hashedPayload}";
        var credentialScope = $"{date}/{service}/tc3_request";
        var stringToSign = $"TC3-HMAC-SHA256\n{timestamp}\n{credentialScope}\n{Sha256Hex(canonicalRequest)}";
        var secretDate = HmacSha256(Encoding.UTF8.GetBytes("TC3" + settings.TencentSecretKey.Trim()), date);
        var secretService = HmacSha256(secretDate, service);
        var secretSigning = HmacSha256(secretService, "tc3_request");
        var signature = ToHex(HmacSha256(secretSigning, stringToSign));
        var authorization = $"TC3-HMAC-SHA256 Credential={settings.TencentSecretId.Trim()}/{credentialScope}, SignedHeaders={signedHeaders}, Signature={signature}";

        using var request = new HttpRequestMessage(HttpMethod.Post, $"https://{host}");
        request.Content = new StringContent(payload, Encoding.UTF8, "application/json");
        request.Content.Headers.ContentType!.CharSet = "utf-8";
        request.Headers.TryAddWithoutValidation("Authorization", authorization);
        request.Headers.TryAddWithoutValidation("Host", host);
        request.Headers.TryAddWithoutValidation("X-TC-Action", action);
        request.Headers.TryAddWithoutValidation("X-TC-Timestamp", timestamp.ToString(CultureInfo.InvariantCulture));
        request.Headers.TryAddWithoutValidation("X-TC-Version", version);
        request.Headers.TryAddWithoutValidation("X-TC-Region", region);
        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var text = await response.Content.ReadAsStringAsync(cancellationToken);
        EnsureSuccess(response, text);
        using var document = JsonDocument.Parse(text);
        var root = document.RootElement.GetProperty("Response");
        if (root.TryGetProperty("Error", out var error))
        {
            throw new InvalidOperationException(error.GetProperty("Message").GetString() ?? "腾讯翻译失败。");
        }

        return root.GetProperty("TargetText").GetString() ?? "";
    }

    private static bool LooksLikeHtml(string value)
    {
        return value.Contains('<') && value.Contains('>') && HtmlTagRegex.IsMatch(value);
    }

    private static StringContent JsonContent(object value)
    {
        return new StringContent(JsonSerializer.Serialize(value, JsonOptions), Encoding.UTF8, "application/json");
    }

    private static void EnsureSuccess(HttpResponseMessage response, string text)
    {
        if (response.IsSuccessStatusCode)
        {
            return;
        }

        throw new InvalidOperationException($"翻译请求失败 HTTP {(int)response.StatusCode}：{text}");
    }

    private static string NormalizeProvider(string? provider)
    {
        return provider?.Trim().ToLowerInvariant() switch
        {
            "baidu" => "baidu",
            "tencent" => "tencent",
            "google" => "google",
            _ => "microsoft"
        };
    }

    private static string LanguageForBaidu(string language)
    {
        return NormalizeLanguage(language) switch
        {
            "en" => "en",
            "ja" => "jp",
            "ko" => "kor",
            "fr" => "fra",
            "es" => "spa",
            "ru" => "ru",
            "de" => "de",
            "zh-hant" => "cht",
            _ => "zh"
        };
    }

    private static string LanguageForTencent(string language)
    {
        return NormalizeLanguage(language) switch
        {
            "zh-hant" => "zh-TW",
            "zh" => "zh",
            var value => value
        };
    }

    private static string LanguageForGoogle(string language)
    {
        return NormalizeLanguage(language) switch
        {
            "zh-hant" => "zh-TW",
            "zh" => "zh-CN",
            var value => value
        };
    }

    private static string LanguageForMicrosoft(string language)
    {
        return NormalizeLanguage(language) switch
        {
            "zh-hant" => "zh-Hant",
            "zh" => "zh-Hans",
            var value => value
        };
    }

    private static string NormalizeLanguage(string language)
    {
        var value = language.Trim().ToLowerInvariant();
        return value switch
        {
            "中文" or "简体中文" or "chinese" or "zh-cn" or "zh_hans" or "zh-hans" => "zh",
            "繁体中文" or "zh-tw" or "zh_hant" or "zh-hant" => "zh-hant",
            "英文" or "英语" or "english" => "en",
            "日文" or "日语" or "japanese" => "ja",
            "韩文" or "韩语" or "korean" => "ko",
            "法文" or "法语" or "french" => "fr",
            "西班牙文" or "西班牙语" or "spanish" => "es",
            "俄文" or "俄语" or "russian" => "ru",
            "德文" or "德语" or "german" => "de",
            _ => string.IsNullOrWhiteSpace(value) ? "zh" : value
        };
    }

    private static string Md5Hex(string value)
    {
        return ToHex(MD5.HashData(Encoding.UTF8.GetBytes(value)));
    }

    private static string Sha256Hex(string value)
    {
        return ToHex(SHA256.HashData(Encoding.UTF8.GetBytes(value)));
    }

    private static byte[] HmacSha256(byte[] key, string value)
    {
        using var hmac = new HMACSHA256(key);
        return hmac.ComputeHash(Encoding.UTF8.GetBytes(value));
    }

    private static string ToHex(byte[] bytes)
    {
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}

File: android/MRSS/app/src/main/java/com/mrss/app/network/StandardTranslationClient.java
package com.mrss.app.network;

import com.mrss.app.model.ArticleTranslation;
import com.mrss.app.model.StandardTranslationSettings;
import com.mrss.app.model.TranslationJob;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class StandardTranslationClient {
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final MediaType FORM = MediaType.get("application/x-www-form-urlencoded; charset=utf-8");
    private static final Pattern HTML_TAG = Pattern.compile("(?s)<[^>]+>");
    private final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(45, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .callTimeout(60, TimeUnit.SECONDS)
            .build();

    public ArticleTranslation translate(StandardTranslationSettings settings, TranslationJob job) throws Exception {
        if (settings == null) {
            throw new IllegalStateException("常规翻译设置不存在");
        }
        ArticleTranslation result = new ArticleTranslation();
        result.title = translatePlain(settings, firstNonEmpty(job.title, ""), job.targetLanguage);
        result.content = translateKeepingFormat(settings, firstNonEmpty(job.content, ""), job.targetLanguage);
        return result;
    }

    private String translateKeepingFormat(StandardTranslationSettings settings, String content, String targetLanguage) throws Exception {
        if (isBlank(content)) {
            return "";
        }
        List<String> textParts = new ArrayList<>();
        List<String> tags = new ArrayList<>();
        Matcher matcher = HTML_TAG.matcher(content);
        int last = 0;
        while (matcher.find()) {
            textParts.add(content.substring(last, matcher.start()));
            tags.add(matcher.group());
            last = matcher.end();
        }
        textParts.add(content.substring(last));
        if (tags.isEmpty()) {
            return translatePlain(settings, content, targetLanguage);
        }
        StringBuilder translated = new StringBuilder();
        for (int i = 0; i < textParts.size(); i++) {
            translated.append(translatePlain(settings, textParts.get(i), targetLanguage));
            if (i < tags.size()) {
                translated.append(tags.get(i));
            }
        }
        return translated.toString();
    }

    private String translatePlain(StandardTranslationSettings settings, String text, String targetLanguage) throws Exception {
        if (isBlank(text) || text.trim().matches("^[\\s\\p{Punct}]+$")) {
            return text;
        }
        String provider = normalizeProvider(settings.provider);
        if ("baidu".equals(provider)) {
            return translateBaidu(settings, text, targetLanguage);
        }
        if ("tencent".equals(provider)) {
            return translateTencent(settings, text, targetLanguage);
        }
        if ("google".equals(provider)) {
            return translateGoogle(settings, text, targetLanguage);
        }
        return translateMicrosoft(settings, text, targetLanguage);
    }

    private String translateBaidu(StandardTranslationSettings settings, String text, String targetLanguage) throws Exception {
        require(settings.baiduAppId, "请填写百度翻译 App ID");
        require(settings.baiduSecret, "请填写百度翻译密钥");
        String salt = String.valueOf(System.currentTimeMillis());
        String to = languageCode(targetLanguage, "baidu");
        String sign = md5(settings.baiduAppId + text + salt + settings.baiduSecret);
        String body = "q=" + encode(text)
                + "&from=auto"
                + "&to=" + encode(to)
                + "&appid=" + encode(settings.baiduAppId)
                + "&salt=" + encode(salt)
                + "&sign=" + encode(sign);
        String response = post("https://fanyi-api.baidu.com/api/trans/vip/translate", body, FORM, null);
        JSONObject object = new JSONObject(response);
        if (object.has("error_code")) {
            throw new IllegalStateException("百度翻译失败 " + object.optString("error_code") + "：" + object.optString("error_msg"));
        }
        JSONArray array = object.getJSONArray("trans_result");
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < array.length(); i++) {
            if (i > 0) {
                builder.append('\n');
            }
            builder.append(array.getJSONObject(i).optString("dst"));
        }
        return builder.toString();
    }

    private String translateMicrosoft(StandardTranslationSettings settings, String text, String targetLanguage) throws Exception {
        require(settings.microsoftKey, "请填写微软翻译 Key");
        String to = languageCode(targetLanguage, "microsoft");
        JSONArray body = new JSONArray().put(new JSONObject().put("Text", text));
        Request.Builder builder = new Request.Builder()
                .url("https://api.cognitive.microsofttranslator.com/translate?api-version=3.0&to=" + encode(to))
                .post(RequestBody.create(body.toString(), JSON))
                .header("Content-Type", "application/json; charset=utf-8")
                .header("Ocp-Apim-Subscription-Key", settings.microsoftKey.trim());
        if (!isBlank(settings.microsoftRegion) && !"global".equalsIgnoreCase(settings.microsoftRegion.trim())) {
            builder.header("Ocp-Apim-Subscription-Region", settings.microsoftRegion.trim());
        }
        String response = execute(builder.build());
        return new JSONArray(response).getJSONObject(0).getJSONArray("translations").getJSONObject(0).optString("text");
    }

    private String translateGoogle(StandardTranslationSettings settings, String text, String targetLanguage) throws Exception {
        require(settings.googleApiKey, "请填写 Google 翻译 API Key");
        JSONObject body = new JSONObject()
                .put("q", new JSONArray().put(text))
                .put("target", languageCode(targetLanguage, "google"))
                .put("format", "text");
        String response = post("https://translation.googleapis.com/language/translate/v2?key=" + encode(settings.googleApiKey), body.toString(), JSON, null);
        return new JSONObject(response).getJSONObject("data").getJSONArray("translations").getJSONObject(0).optString("translatedText");
    }

    private String translateTencent(StandardTranslationSettings settings, String text, String targetLanguage) throws Exception {
        require(settings.tencentSecretId, "请填写腾讯云 SecretId");
        require(settings.tencentSecretKey, "请填写腾讯云 SecretKey");
        String host = "tmt.tencentcloudapi.com";
        String service = "tmt";
        long timestamp = System.currentTimeMillis() / 1000L;
        String payload = new JSONObject()
                .put("SourceText", text)
                .put("Source", "auto")
                .put("Target", languageCode(targetLanguage, "tencent"))
                .put("ProjectId", 0)
                .toString();
        String authorization = tencentAuthorization(settings.tencentSecretId.trim(), settings.tencentSecretKey.trim(), host, service, timestamp, payload);
        Request request = new Request.Builder()
                .url("https://" + host)
                .post(RequestBody.create(payload, JSON))
                .header("Authorization", authorization)
                .header("Content-Type", "application/json; charset=utf-8")
                .header("Host", host)
                .header("X-TC-Action", "TextTranslate")
                .header("X-TC-Version", "2018-03-21")
                .header("X-TC-Region", isBlank(settings.tencentRegion) ? "ap-beijing" : settings.tencentRegion.trim())
                .header("X-TC-Timestamp", String.valueOf(timestamp))
                .build();
        String response = execute(request);
        JSONObject object = new JSONObject(response).getJSONObject("Response");
        if (object.has("Error")) {
            JSONObject error = object.getJSONObject("Error");
            throw new IllegalStateException("腾讯翻译失败 " + error.optString("Code") + "：" + error.optString("Message"));
        }
        return object.optString("TargetText");
    }

    private String tencentAuthorization(String secretId, String secretKey, String host, String service, long timestamp, String payload) throws Exception {
        String algorithm = "TC3-HMAC-SHA256";
        String date = utcDate(timestamp);
        String canonicalHeaders = "content-type:application/json; charset=utf-8\nhost:" + host + "\n";
        String signedHeaders = "content-type;host";
        String canonicalRequest = "POST\n/\n\n" + canonicalHeaders + "\n" + signedHeaders + "\n" + sha256Hex(payload);
        String credentialScope = date + "/" + service + "/tc3_request";
        String stringToSign = algorithm + "\n" + timestamp + "\n" + credentialScope + "\n" + sha256Hex(canonicalRequest);
        byte[] secretDate = hmac256(("TC3" + secretKey).getBytes(StandardCharsets.UTF_8), date);
        byte[] secretService = hmac256(secretDate, service);
        byte[] secretSigning = hmac256(secretService, "tc3_request");
        String signature = bytesToHex(hmac256(secretSigning, stringToSign));
        return algorithm + " Credential=" + secretId + "/" + credentialScope + ", SignedHeaders=" + signedHeaders + ", Signature=" + signature;
    }

    private String post(String url, String body, MediaType mediaType, Request.Builder headers) throws Exception {
        Request.Builder builder = headers == null ? new Request.Builder() : headers;
        return execute(builder.url(url).post(RequestBody.create(body, mediaType)).build());
    }

    private String execute(Request request) throws Exception {
        try (Response response = client.newCall(request).execute()) {
            String text = response.body() == null ? "" : response.body().string();
            if (!response.isSuccessful()) {
                throw new IllegalStateException("翻译请求失败 " + response.code() + "：" + (isBlank(text) ? "服务器没有返回错误详情" : text));
            }
            return text;
        }
    }

    private String languageCode(String language, String provider) {
        String value = isBlank(language) ? "中文" : language.trim().toLowerCase(Locale.ROOT);
        boolean chinese = value.contains("中文") || value.contains("chinese") || value.equals("zh") || value.equals("zh-cn") || value.equals("zh-hans");
        boolean traditionalChinese = value.contains("繁体") || value.equals("zh-tw") || value.equals("zh-hant");
        boolean english = value.contains("英文") || value.contains("英语") || value.contains("english") || value.equals("en");
        boolean japanese = value.contains("日文") || value.contains("日语") || value.contains("japanese") || value.equals("ja") || value.equals("jp");
        boolean korean = value.contains("韩文") || value.contains("韩语") || value.contains("korean") || value.equals("ko");
        boolean french = value.contains("法文") || value.contains("法语") || value.contains("french") || value.equals("fr");
        boolean spanish = value.contains("西班牙") || value.contains("spanish") || value.equals("es");
        boolean russian = value.contains("俄文") || value.contains("俄语") || value.contains("russian") || value.equals("ru");
        boolean german = value.contains("德文") || value.contains("德语") || value.contains("german") || value.equals("de");
        if ("baidu".equals(provider)) {
            if (traditionalChinese) return "cht";
            if (chinese) return "zh";
            if (english) return "en";
            if (japanese) return "jp";
            if (korean) return "kor";
            if (french) return "fra";
            if (spanish) return "spa";
            if (russian) return "ru";
            if (german) return "de";
        }
        if ("google".equals(provider)) {
            if (traditionalChinese) return "zh-TW";
            if (chinese) return "zh-CN";
        }
        if ("microsoft".equals(provider)) {
            if (traditionalChinese) return "zh-Hant";
            if (chinese) return "zh-Hans";
        }
        if ("tencent".equals(provider)) {
            if (traditionalChinese) return "zh-TW";
            if (chinese) return "zh";
        }
        if (chinese) return "zh";
        if (english) return "en";
        if (japanese) return "ja";
        if (korean) return "ko";
        if (french) return "fr";
        if (spanish) return "es";
        if (russian) return "ru";
        if (german) return "de";
        return language.trim();
    }

    private String normalizeProvider(String provider) {
        if (isBlank(provider)) {
            return "microsoft";
        }
        String value = provider.trim().toLowerCase(Locale.ROOT);
        if ("baidu".equals(value) || "tencent".equals(value) || "google".equals(value)) {
            return value;
        }
        return "microsoft";
    }

    private void require(String value, String message) {
        if (isBlank(value)) {
            throw new IllegalStateException(message);
        }
    }

    private String encode(String value) throws Exception {
        return URLEncoder.encode(value == null ? "" : value, "UTF-8");
    }

    private String firstNonEmpty(String value, String fallback) {
        return isBlank(value) ? fallback : value;
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    private String md5(String value) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        return bytesToHex(md.digest(value.getBytes(StandardCharsets.UTF_8)));
    }

    private String sha256Hex(String value) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        return bytesToHex(md.digest(value.getBytes(StandardCharsets.UTF_8)));
    }

    private byte[] hmac256(byte[] key, String value) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(key, "HmacSHA256"));
        return mac.doFinal(value.getBytes(StandardCharsets.UTF_8));
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder builder = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            builder.append(String.format(Locale.ROOT, "%02x", b & 0xff));
        }
        return builder.toString();
    }

    private String utcDate(long timestamp) {
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd", Locale.ROOT);
        format.setTimeZone(TimeZone.getTimeZone("UTC"));
        return format.format(new Date(timestamp * 1000L));
    }
}
